lib/OpenQA/WebAPI/Plugin/Helpers.pm

```diff
@@ -364,6 +364,38 @@ sub register {
             );
         });
 
+    $app->helper(
+        populate_hash_with_needle_timestamps_and_urls => sub {
+            my ($c, $needle, $hash) = @_;
+
+            $hash->{last_seen}  = $needle ? $needle->last_seen_time    || 'never' : 'unknown';
+            $hash->{last_match} = $needle ? $needle->last_matched_time || 'never' : 'unknown';
+            return $hash unless $needle;
+            if (my $last_seen_module_id = $needle->last_seen_module_id) {
+                $hash->{last_seen_link} = $c->url_for(
+                    'admin_needle_module',
+                    module_id => $last_seen_module_id,
+                    needle_id => $needle->id
+                );
+            }
+            if (my $last_matched_module_id = $needle->last_matched_module_id) {
+                $hash->{last_match_link} = $c->url_for(
+                    'admin_needle_module',
+                    module_id => $last_matched_module_id,
+                    needle_id => $needle->id
+                );
+            }
+            return $hash;
+        });
+
+    $app->helper(
+        popover_link => sub {
+            my ($c, $text, $url) = @_;
+            return $text unless $url;
+            return "<a href='$url'>$text</a>";
+            # note: Not using link_to because the tag is supposed to be escaped.
```

Comment:
You mean the tag is already escaped? Maybe "Not using link_to because it would escape the already escaped $url" would be clearer.

Reply:
The string (containing HTML code) should be escaped (it is not already escaped) and hence I'm returning the bare string instead of using link_to which would not lead to an escaped a tag.

Reply:
So the string - the HTML string built and returned by the function - needs to be escaped? If that is what you mean, I don't understand how that is desirable. It probably doesn't help that [link_to](https://mojolicious.org/perldoc/Mojolicious/Plugin/TagHelpers#link_to) doesn't document how it treats encoding or escaping.

Reply:
> If that is what you mean, I don't understand how that is desirable.

It is desirable because it is rendered within an HTML attribute. That's how these bootstrap popovers work.

I'm just referring to `link_to` specifically because this function makes a link but `link_to` behaves like any other helper.